docs/specs/3-13-2023-commands-fit-n-finsh.md

```diff
@@ -0,0 +1,71 @@
+# 3/13/2023 - Commands Fit-and-Finish Proposal
+
+## Design
+
+Commands that interact with the core development workflow will be verb-based. This feels natural to the user as the user is building an application, provisioning infrastructure, and deploying their application to Azure -- which are all actions.
+
+For commands that are of configuration and management in-nature, it is fine to add a noun for a subcommand. For example, `azd env <action>` and `azd config <action>`.
+
+This means that the overall proposed design matches with what `azd` currently offers, and no changes need to occur.
+
+## `azd` usage output
+
+To help introduce user to the command layout, I propose slight changes to the current `azd` usage output (with UX input needed), that looks like the following:
+
+```
+Commands
+  Develop an app
+    init     	: Initialize a new application.
+    provision	: Provision the Azure resources for an application.
+    deploy   	: Deploy the application's code to Azure.
+    up       	: Initialize application, provision Azure resources, and deploy your project with a single command.
+
+  Monitor, test, and release your app
+    monitor  	: Monitor a deployed application.
+    pipeline 	: Manage and configure your deployment pipelines.
+
+  Authenticate with Azure
+    login    	: Log in to Azure.
+    logout   	: Log out of Azure.
+
+  Miscellaneous commands
+    config   	: Manage azd configurations (ex: default Azure subscription, location).
+    down     	: Delete Azure resources for an application.
+    env      	: Manage environments.
```

Comment:
restore sits at the same spot as the build command for me. If we don't need to expose build, I don't know why we need to expose restore. Up would automatically restore, no? We made the argument that build isn't needed because people can go to each of their projects and restore. I feel like we should at least be consistent for GA. restore and build across all projects in azure yaml are both inner loop conveniences to span across languages and projects.

Reply:
Whether or not `azd build` was exposed was based on a timing issue for GA.  I'm making great progress so far and don't see any issues that would be blocking to expose `build` as part of GA.  I'd also vote to keep the granular `restore` command which will join `package` and `publish` as they become available.

Reply:
I think the thing I struggle with here is that we have all of these dev workflow commands ("restore", soon "build" (and maybe some more like `package` and `publish`) as well as new ones like `clean`) and I'm not sure who they are in service of.

I'm never going to put `azd build` on my inner loop.  If I'm writing `go` code, my fingers are wired to use `go install`.  If I'm doing dotnet stuff, typing `dotnet run` or hitting F5 in VS.  Some of these dev gestures don't make sense across all languages (I wonder what a `azd clean` is going to do for every language we're going to support).  We already have this problem today with `restore` where we had to provide a semantics for `restore` when you are using Docker and the semantics, we gave it have nothing to do with docker.

I think the place that this helps is the ability for me to walk up to a project I know nothing about, in a language/toolchain I have no familiarity with, and just say `azd build` anywhere in the system and then everything gets built.  Or I can run `azd restore` and everything gets restored.  But I don't understand why I would want to do these things and why the tool I'm reaching for to do these things `azd`.

For `package` and `publish` I can tell myself a story as to why these make sense as exposed commands.  There is some "azure specificness" baked into both of them (for `package` the thing we produce depends on the service we are going to deploy to, and for `publish` the action is service specific.)  The rest of these, it seems like there's no tie to Azure at all and so I wonder why I have to use an azure specific tool to do them or what value I get out of doing so.

Reply:
+1

I am also trying to understand how much value we provide with such build life-cycle commands which are just wrappers calling into language specific build commands. This is a slippery slope. Once we get into exposing some parts of it. We will be forced to keep adding to it. Also, every language has its own way and abstractions in their build lifecycle. We adopting them and having a common way to expose things would be too much to tackle. 

Instead, let developers use the way they are with VSCode/other IDE's and their build tools etc. Even for the local dev loop without Azd. I believe there should be good enough tooling support from VSCode and the language specific tooling. Should we be in the middle and/or be wrappers, duplicating them in Azd for completeness. I believe folks have different perspectives on this.

We should have some agreed upon guiding principles. So that we all are on the same page.

Reply:
As Rajesh and Matt points out, I equally have a concern with inner-loop workflow commands. The user can definitely have their own way setup for restore/build.

If we do introduce this, I think we'd be forced to also introduce other useful workflows such as `build`. The problem I have is that it simply adds unnecessary concepts to `azd`, and surfaces unnecessary gestures. The proposition of `azd` is to deploy to Azure. If we start doing dev workflows, users might ask for more. Users might ask for `azd lint` to support linting their source code.

And thus, I'd rather us caution on the side of safety and only introduce this post-GA when we hear clear user feedback.

Reply:
For restore specifically, it doesn't look like we have a lot of [GitHub usage](https://github.com/search?q=%22azd+restore%22+language%3AJSON&type=code&l=JSON) in `tasks.json` aside from `todo` and copying directly from `todo`.

Reply:
With all of the above comments, I think we need to be careful about becoming too granular. It's a slippery slope and being high-level is a central tenet (and major value prop) of azd. I really worry about exposing all of these commands (e.g. build, package, publish, lint, clean, etc) and having them bury the very simple core workflow we have today.

I think something we should all consider is how to have azd play nicely with the rest of the developer's toolchain, without trying to replace it all the pieces of it.

That's not to say that these commands wouldn't be useful but thinking more about this, perhaps we should avoid exposing commands where tools in the developer's current workflow  (e.g. pip, npm) can already handle a particular task for them. azd will never be the only CLI they use for developing their application.

Reply:
I've captured all the thoughts and suggestions in `#1491` in our private repo. Everyone is tagged in it.